### PR TITLE
Drop sync InvokeOnPartitions, use IOPAsync.get()

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java [hunk 1]
```diff
@@ -16,137 +16,212 @@
...
-        awaitCompletion();
-
-        retryFailedPartitions();
-
-        return (Map<Integer, T>) partitionResults;
+        return future;
     }
 
     private void ensureNotCallingFromPartitionOperationThread() {
+        // TODO [viliam] is this check necessary when it's async?
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: Probably this check can be moved to a subclass of `SimpleCompletableFuture` with overridden `get(long, java.util.concurrent.TimeUnit)`.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java [hunk 2]
```diff
@@ -16,137 +16,212 @@
...
             List<Integer> partitions = mp.getValue();
             PartitionIteratingOperation op = new PartitionIteratingOperation(operationFactory, toIntArray(partitions));
-            Future future = operationService.createInvocationBuilder(serviceName, op, address)
+            operationService.createInvocationBuilder(serviceName, op, address)
                     .setTryCount(TRY_COUNT)
                     .setTryPauseMillis(TRY_PAUSE_MILLIS)
-                    .invoke();
-            futures.put(address, future);
+                    .invoke()
+                    .andThen(new FirstAttemptExecutionCallback(mp.getValue()));
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: Not a big deal, but `mp.getValue()` is already fetched into `partitions`.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java [hunk 3]
```diff
@@ -16,137 +16,212 @@
...
     <T> Map<Integer, T> invoke() throws Exception {
-        ensureNotCallingFromPartitionOperationThread();
+        return this.<T>invokeAsync(null).get();
+    }
 
+    /**
+     * Executes all the operations on the partitions.
+     */
+    @SuppressWarnings("unchecked")
+    <T> ICompletableFuture<Map<Integer, T>> invokeAsync(ExecutionCallback<Map<Integer, T>> callback) {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: Do we really need this additional `callback` parameter? Looks like the same behaviour can be handled by simple `andThen` on the caller side. Additionally, we will avoid ambiguity between the order of this callback and extra `andThen`s the caller might have installed (now the callback is executed after all `andThen`s).

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java [hunk 4]
```diff
@@ -16,137 +16,212 @@
...
+
+                            @Override
+                            public void onFailure(Throwable t) {
+                                setPartitionResult(partitionId, t);
+                                decrementLatchAndHandle(1);
+                            }
+                        });
+    }
+
+    private void decrementLatchAndHandle(int count) {
```
Comment (taburet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: To me it feels like an error-prone situation when we return results for lower number of partitions than requested. It's unclear what such situations indicate: we skipped/forgot some partitions entirely or we processed them without results. We already had such problematic cases recently, EPs were skipping partitions silently and it was really hard to diagnose/debug this.

Not sure what we should do, at least some kind of a logging here and in `FirstAttemptExecutionCallback.onResponse` would be really helpful; a more safe option is to completely disallow such situations by reporting an exceptional result to the future.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java (hunk 3 above)
Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: I've got similar comment several times. My response is: some async operations use callback, others use the returned future. Of course, they are replacable by each other. But why convert the api?

For example, `IMap.submitToKey` is available in both flavors, async version of which I have been creating. I don't know why is this. Tell me which one I should keep and I'll do it.

Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: The problem is that one may use `invokeAsync` in those two ways simultaneously and they are slightly conflicting, `submitToKey` is either accepting a callback or returning a future. Not a big deal for me anyway :)

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java (hunk 4 above)
Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: This behavior was allowed by the old version. I try to make `invokeAsync().get()` equivalent to the old version, that's why I allowed it.

Anyway, I added more checks, as it seems to me that `PartitionIteratingOperation` always returns responses for all partitions.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java [hunk 5]
```diff
@@ -1144,7 +1144,7 @@ private static void validateEntryProcessorForSingleKeyProcessing(EntryProcessor
      * Async version of {@link #executeOnKeysInternal}.
      */
     public ICompletableFuture<Map<K, Object>> executeOnKeysInternalAsync(Set<K> keys, Set<Data> dataKeys,
-                EntryProcessor entryProcessor, final ExecutionCallback<Map<K, Object>> callback) {
+                EntryProcessor entryProcessor) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java]: Can we now make `exeucteOnKeysInternal` forward to this method and get rid of the `TODO`s in lines 1117-1120?

**Comment by vbekiaris** [comment]
👍

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java (hunk 1 above)
Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java]: As per discussion we're leaving it here and remove the TODO.

**Pull request merged**